ott/core/initializers.py

```diff
@@ -0,0 +1,330 @@
+    tol: DualSort convergence threshold. Defaults to 1e-2.
+    max_iter: Max DualSort steps. Defaults to 100.
+    stop_gradient: Do not trace gradient. Defaults to True.
+  """
+
+  def __init__(
+      self,
+      vector_min: bool = True,
+      tol: float = 1e-2,
+      max_iter: int = 100,
+      stop_gradient: bool = True
+  ) -> None:
+
+    super().__init__()
+
+    self.tolerance = tol
+    self.stop_gradient = stop_gradient
+    self.max_iter = max_iter
+    self.update_fn = lambda f, mod_cost: jax.lax.cond(
+        vector_min, self.vectorized_update, self.coordinate_update, f, mod_cost
+    )
+
+  def vectorized_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Inner loop DualSort Update.
+
+    Args:
+      f : potential f, array of size n.
+      modified_cost: cost matrix minus diagonal column-wise.
+
+    Returns:
+      updated potential vector, f.
+    """
+    f = jnp.min(modified_cost + f[None, :], axis=1)
+    return f
+
+  def coordinate_update(
+      self, f: jnp.ndarray, modified_cost: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Coordinate-wise updates within inner loop.
+
+    Args:
+      f: potential f, array of size n.
+      modified_cost: cost matrix minus diagonal column-wise.
+
+    Returns:
+      updated potential vector, f.
+    """
+
+    def body_fn(i, f):
+      new_f = jnp.min(modified_cost[i, :] + f)
+      f = f.at[i].set(new_f)
+      return f
+
+    return jax.lax.fori_loop(0, len(f), body_fn, f)
+
+  def init_sorting_dual(
+      self, modified_cost: jnp.ndarray, f_potential: jnp.ndarray
+  ) -> jnp.ndarray:
+    """Run DualSort algorithm.
+
+    Args:
+      modified_cost:  cost matrix minus diagonal column-wise.
+      f_potential: potential f, array of size n.
+
+    Returns:
+      potential f, array of size n.
+    """
+
+    def body_fn(state):
+      prev_f, _, it = state
+      f_potential = self.update_fn(prev_f, modified_cost)
+      diff = jnp.sum((f_potential - prev_f) ** 2)
+      it += 1
+      return f_potential, diff, it
+
+    def cond_fn(state):
+      _, diff, it = state
+      return jnp.logical_and(diff > self.tolerance, it < self.max_iter)
+
+    it = 0
+    diff = self.tolerance + 1.0
+    state = (f_potential, diff, it)
+
+    f_potential, _, it = jax.lax.while_loop(
+        cond_fun=cond_fn, body_fun=body_fn, init_val=state
+    )
+
+    return f_potential
+
+  def init_dual_a(
+      self,
+      ot_problem: linear_problems.LinearProblem,
+      lse_mode: bool,
+      init_f: Optional[jnp.ndarray] = None,
+  ) -> jnp.ndarray:
+    """Apply DualSort algo.
+
+    Args:
+      ot_problem: OT problem.
+      lse_mode: Return potential if true, scaling if false.
+      init_f: potential f, array of size n.
+
+    Returns:
+      potential f, array of size n.
+    """
+    cost_matrix = ot_problem.geom.cost_matrix
+    if self.stop_gradient:
+      cost_matrix = jax.lax.stop_gradient(cost_matrix)
+
```

Comment:
This uses the cost matrix from geom. I assume the tests should go when the cost matrix is created or passed rather than everywhere that touches a cost matrix?

Reply:
Yes, wouldn't add this check everywhere where you need to use the cost, just on the API boundary to catch it early as possible (in this cse `init_dual_a`).

Reply:
Shouldn't it be way before this when initializing the geom object?

Reply:
> Shouldn't it be way before this when initializing the geom object?


Sorry for late reply. If I understand correctly, don't think this should be in `__init__` method of the `Geometry`, since we don't require cost matrix to be square.